src/client/workflow/workflow.ts

```diff
@@ -0,0 +1,262 @@
+   *    will be executed
+   * 3. Called with full results: After the final partial result call returns to Qstash,
+   *    qstash will call again with the full result. In this case, parallel step will
+   *    return the result and
+   */
+  protected getParallelCallState(parallelStepCount: number): PARALLEL_CALL_STATE {
+    const steps = this.steps.filter(
+      (step) => (step.stepId === 0 ? step.targetStep : step.stepId) >= this.stepCount
+    );
+
+    if (steps.length === 0) {
+      return "first";
+      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+    } else if (steps.length >= 2 * parallelStepCount) {
+      return "last";
+    } else if (steps.at(-1)?.targetStep) {
+      return "partial";
+    } else {
+      return "discard";
+    }
+  }
+
+  private addResult(result: unknown, stepId?: number) {
+    this.addStep({
+      stepId: stepId ?? this.stepCount,
+      out: result,
+      concurrent: 1,
+      targetStep: 0,
+    });
+  }
+
+  private addStep(step: Step | Step[]) {
+    if (Array.isArray(step)) {
+      this.pendingSteps = [...this.pendingSteps, ...step];
+    } else {
+      this.pendingSteps.push(step);
+    }
+  }
+
+  private async invokeQstash(step: Step) {
+    const headers: Record<string, string> = {
+      [`Upstash-Forward-${internalHeader}`]: "yes",
+      "Upstash-Forward-Upstash-Workflow-Id": this.workflowId,
+      "Upstash-Workflow-Id": this.workflowId,
+    };
+
+    await this.client.publishJSON({
+      headers: headers,
+      method: "POST",
+      body: JSON.stringify(step),
+      url: this.url,
+      notBefore: step.sleepUntil,
+      delay: step.sleepFor,
+    });
+  }
+
+  private async sendPendingToQstash() {
+    // TODO: batch request for concurrent requests
+    for (const step of this.pendingSteps) {
+      await this.invokeQstash(step);
+    }
+    this.pendingSteps = [];
+  }
+
+  /**
+   * STATIC METHODS
+   */
+
+  static async parseRequest(request: Request) {
+    const callHeader = request.headers.get(internalHeader);
+    const firstCall = !callHeader;
+
+    const workflowId = firstCall ? `wf${nanoid()}` : request.headers.get(workflowIdHeader) ?? "";
+    if (workflowId.length === 0) {
+      throw new Error("Couldn't get workflow id from header");
+    }
+
+    let payload: string[] | undefined;
+    try {
+      payload = (await request.json()) as string[];
+    } catch {
+      payload = undefined;
+    }
+
+    let steps: Step[];
+    if (firstCall) {
+      steps = [
+        {
+          stepId: 0,
+          out: payload,
+          concurrent: 1,
+          targetStep: 0,
+        },
+      ];
+    } else {
+      if (payload === undefined) {
+        throw new Error("only first call can have an empty body");
+      }
+
+      steps = payload.map((rawStep) => {
+        return JSON.parse(JSON.parse(Buffer.from(rawStep, "base64").toString()) as string) as Step;
+      });
+    }
```

Comment:
```ts
const payload = await WorkflowParser.parsePayload(request, isFirstStep);
const steps = WorkflowParser.generateSteps(payload, isFirstStep);
```

Maybe we can break this up like this instead of a big single function.